### DISCUSSION TITLE
Import and export option for flows

### DISCUSSION THREAD

**Discussion by MikeVanDeeAvrotros** [directus/directus#18618] [Feature Requests]
### Summary

It would be nice to have an import and export option for flows like the other content models. 
![image](https://github.com/directus/directus/assets/32135919/6e3e146d-815a-48f3-857e-58589363a3ae)

### Basic Example

Using the [import](https://docs.directus.io/reference/system/utilities.html#export-data-to-a-file) and [export](https://docs.directus.io/reference/system/utilities.html#export-data-to-a-file) API endpoints we can import and export flows. like
```http
POST /utils/import/flows

{
 "format": "json"
}

POST /utils/export/flows

Content-Type: multipart/form-data; charset=utf-8; boundary=__X_BOUNDARY__
...
```

### Motivation

- This would make reusing flows between Directus instances easier. 
- No need for database exports and imports from developers. 
- Constancy between different parts of the admin app.

### Detailed Design

The [import](https://docs.directus.io/reference/system/utilities.html#export-data-to-a-file) and [export](https://docs.directus.io/reference/system/utilities.html#export-data-to-a-file) API endpoints would need support for `flows` data model. 

```http
POST /utils/import/flows

POST /utils/export/flows
```

### Must Have:
- Import flows
- export flows

### Should Have:
- Row selection for flows like content items. 

### Could Have:
- Field filters

### Won't Have:
- Layout options like content

### Drawbacks

I'm not familiar enough with the Directus core to make good assessment of drawbacks, it uses existing components. However if the components are not compatible with flows model it can be more work than expected. 

### Alternatives

There are other alternatives. Writing and extension with this feature. Or manual database exports and imports.

### Adoption Strategy

It should be an easy adoption for developers and users. It uses existing components and is located where you expect it to be.

### Unresolved Questions

- Is import/export compatible with flows?

**Comment by M-GRV** [comment]
Just commented about this yesterday lol :p yes it does indeed need a UI element 

https://github.com/directus/directus/discussions/14405#discussioncomment-5906348

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by rijkvanzanten** [comment]
Hi all! This has unexpectedly progressed quite a lot recently! @ConnorSimply has been doing some great work improving Insights, including the ability to export/import dashboards and realized it greatly reduced the overhead of supporting duplicating flows as well. @ConnorSimply mind posting your thoughts on what the remaining unresolved questions were for enabling duplication on flows? Thanks! 🙂

**Reply by connorwinston** [reply]
The implementation of importing and exporting dashboards is done app side in the QoL PR since they are app only features, it would be fairly straight forward to implement with flows right now...however this wouldn't be an API change it would be an app side change. This request is for adding importing to the API for flows, not changing the APP.

Importing flows through the API is **currently supported and possible**, it just requires you to structure the data the right way through the resolve and reject fields nesting the related data correctly.

**Reply by JayShoe** [reply]
Hello ConnorSimply, can you please explain how to export/import flows via the API? Is there a command to do this or is it a manual process?

My 2 cents is that flows should be exportable and importable via something similar to snapshots, but this might not be the place to discuss this.

**Reply by adussarps** [reply]
```
class DirectusFlowTransfer {
  constructor(sourceUrl, sourceEmail, sourcePassword, destinationUrl, destinationEmail, destinationPassword) {
    this.sourceDirectus = new Directus(sourceUrl);
    this.sourceEmail = sourceEmail;
    this.sourcePassword = sourcePassword;

    this.destinationDirectus = new Directus(destinationUrl);
    this.destinationEmail = destinationEmail;
    this.destinationPassword = destinationPassword;
  }

  async loginToSource() {
    await this.sourceDirectus.auth.login({
      email: this.sourceEmail,
      password: this.sourcePassword,
    });
  }

  async loginToDestination() {
    await this.destinationDirectus.auth.login({
      email: this.destinationEmail,
      password: this.destinationPassword,
    });
  }

  async copyFlows() {
    await this.loginToSource();
    await this.loginToDestination();

    const flows = await this.sourceDirectus.items("directus_flows").readByQuery({
      limit: -1,
      fields: [
        "name",
        "description",
        "trigger",
        "accountability",
        "options",
        "operation",
        "operations.*"
      ]
    });

    console.log("Got flows -- " + flows.data.length);

    for (const flow of flows.data) {
      // Check if flow already exists in the destination instance
      const existingFlow = await this.destinationDirectus.items("directus_flows").readByQuery({
        filter: {
          name: {
            _eq: flow.name,
          },
        },
      });

      if (existingFlow.data && existingFlow.data.length > 0) {
        console.log(`Flow '${flow.name}' already exists in the destination instance.`);
      } else {
        // Create the flow in the destination instance
        // Remove date_created and user_created
        for (const operation of flow.operations) {
          delete operation.date_created;
          delete operation.user_created;
          delete operation.resolve;
          delete operation.reject;
          // Add more fields if needed
        }
        await this.destinationDirectus.items("directus_flows").createOne(flow);
        console.log(`Flow '${flow.name}' copied to the destination instance.`);
      }
    }
  }
}
```

Hello there, I stumble upon this thread, here is a little class that I use to copy flows through the sdk. I did not bother to do a well "resolve" "reject" logic though, so you need to redo the connections.

**Comment by NilsBaumgartner1994** [comment]
This is definetly something which we need.

**Comment by w0kyj** [comment]
Request Review: Aug 15, 2024

## Considerations
- export is the easier part of the solution
- import has complexities and validation concerns
    - schema references need to match
    - operations (custom/marketplace) need to exist
    - variables, env, ...
    - should the default import disable the imported flows
    - how to handle updates to existing flows
        - FKs handling between Flow & Operations
            - easiest is to delete all operations and recreate them, however, if flows active in production for example, this could cause interruptions or issues
        - Possible to only merge changes?
    - Flow triggers can contain "collection item dropdown" interfaces. These might not be valid after import if the database schema is different.
    - Same for operations referencing tables that don't exist in the new schema.

## Architecture Considerations
    - configuration as code --> would help with this where the configurations could be managed in repository
    - search / filter and selective import/export
    - current data structure of Flows --> Operations relational data
        - Why? when developed JSON was not well supported by SQL database vendors)
        - now that JSON is better supported, could implement a breaking schema change to store flows as JSON
        - Blob storage, filtering, field returns -- must be considered
            - json filtering must be supported for this to work
    - Add logs for imports that indicate errors, warnings, ...
    - Consider having a default project/flows settings collection (can just be added to the data model already, but might be a nice default to include)
    - Flow Environment variables -- allows for values to be defined and used within the flow
        - encrypted values (not hashed, but reusable)
    - Security / Flow management restrictions
        - Different admin levels for this would be cool - Flow admin that update flow logic, but Full admin that can define settings, permissions, etc



From Chat:
For a user story:
We use tons flows in production and development, across two applications, with the following workflow:
- Run the whole application locally. Modify flows, code, extensions, etc.
- Export the schema, flows, collections, fields, etc to JSON using bcc-code's schema-sync extension.
- Commit this schema to version control and build images.
- In production, import the schema, collections, and flows as defined in the image.

The result is that flows are always consistent between development and production, and the production flows are immutable. (Except that the import regularly breaks) 

What I usually try to do is to keep flows clean; only operations with values that should be same everywhere.
If I need to use different data/values between environments, I use single type collections for the data.

**Comment by JayShoe** [comment]
This exists in an extension. 
https://github.com/baguse/directus-extension-flow-manager
